### ISSUE TITLE
build a new tool or subcommand similar to asciinema2gif

### ISSUE THREAD

**Issue by piaoger** [LegNeato/asciinema-rs#36]
I might be a good idea to build a new tool or subcommand  which convert cast to gif.
asciinema2gif: [link](https://github.com/asciinema/asciicast2gif)

**Comment by LegNeato** [comment]
I think it could go in this tool. I see two logical entry points:

1.  For new recordings: `asciinema rec /path/to/file.cast --format gif/svg/whatever`
2. For existing recordings: `asciinema convert /path/to/file.cast --format gif/svg/whatever`

**Comment by LegNeato** [comment]
Some quick notes after investigation:

* I think the best way forward would be to use https://github.com/jwilm/alacritty to parse and render the tty. There appears to be a way to get their display grid (https://github.com/jwilm/alacritty/blob/master/tests/ref.rs).
* We could go from the alacritty display grid to our own render, but seems hard.
* alacritty uses `glutin` under the covers (https://github.com/jwilm/alacritty/blob/master/src/window.rs).
* Glutin has a headless renderer (https://docs.rs/glutin/0.13.1/glutin/struct.HeadlessRendererBuilder.html).
* An example of how to save the GL data to an image is at https://github.com/bwasty/gltf-viewer/blob/14342da1ae8ba6d832f0088a92743e8a3a428329/src/viewer.rs#L328 and how to use a headless context is at https://github.com/bwasty/gltf-viewer/blob/14342da1ae8ba6d832f0088a92743e8a3a428329/src/viewer.rs#L86.
* Looks like there may be a bug with the headless glutin context on linux: https://github.com/tomaka/glutin/issues/988

**Comment by LegNeato** [comment]
I've been poking at this and have been able to render a static image using `alacritty`. There are some outstanding issues but nothing that can't be resolved, so this is definitely going to be possible somewhat soon.

![output](https://user-images.githubusercontent.com/368904/38265640-51bebc46-372b-11e8-9dbf-5165720faea6.png)

**Comment by LegNeato** [comment]
Update: I got gif support working, including clamping to 60fps and having proper duration between frames to keep file size down:

![output](https://user-images.githubusercontent.com/368904/38284789-ed5b20b8-3771-11e8-8667-307c11f0c13d.gif)

I'm going to clean it up a bit, report my findings to `alacritty` so they can merge their WIP diff, and then we will support gifs!

**Comment by piaoger** [comment]
It seems we can have gif support soon🤔

**Comment by piaoger** [comment]
It seems we can have gif support soon🤔

**Comment by piaoger** [comment]
not sure if gifski is helpful which is gapl licensed.
https://github.com/ImageOptim/gifski

**Comment by ku1ik** [comment]
I've implemented a minimal virtual terminal in Rust for asciinema recently, for the purpose of integrating it into web player (rust-wasm) and screenshot generation: https://github.com/asciinema/vt-rs

It's working nicely, although I haven't had a chance of preparing README, publishing to crates.io, nor integrating into web player. Not sure if useful to you at this stage but I thought mentioning it won't hurt :)

**Comment by LegNeato** [comment]
Sweet, looks awesome! I'll look into integrating, the alacritty changes have stalled.

**Comment by zicklag** [comment]
I've just started working on finding a way to render gif's. I'm posting what I find in a [forum topic](https://users.rust-lang.org/t/suggested-libraries-for-writing-an-asciinema-to-gif-tool/39450/4?u=zicklag).

I think it might be simple enough to support SVG, animated PNG, and Gif with one solution.

I'm going to try to put the converter logic into its own crate lib/cli because I think it would be useful to people who still want to use the normal asciinema CLI, but we can also include the crate as a library to `asciinema-rs` so that it is built-in to this tool as well.

**Comment by piaoger** [comment]
Found someone using this tool:
https://github.com/nbedos/termtosvg

**Comment by zicklag** [comment]
@piaoger Yes, I've used that before. It is very nice. My biggest problem came when I needed to paste a recording into GitHub and they for some reason don't support SVG in comments!

I forgot to post an update by my https://github.com/katharostech/cast2gif/ program is actually basically working. It needs some new features to be a little nicer, but it is functional now. It's a rust crate so we could incorporate it into asciinema-rs.

**Comment by piaoger** [comment]
@zicklag  
yes. I found cast2gif and you also mentioned termtosvg in the post after that.Let me try asciinema-rs and cast2gif together.
